File: cc/jwt/internal/jwt_hmac_key_manager.cc
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
///////////////////////////////////////////////////////////////////////////////

#include "tink/jwt/internal/jwt_hmac_key_manager.h"

#include <map>

#include "absl/strings/string_view.h"
#include "tink/jwt/internal/raw_jwt_hmac_key_manager.h"
#include "tink/mac.h"
#include "tink/subtle/hmac_boringssl.h"
#include "tink/subtle/random.h"
#include "tink/util/enums.h"
#include "tink/util/errors.h"
#include "tink/util/input_stream_util.h"
#include "tink/util/protobuf_helper.h"
#include "tink/util/status.h"
#include "tink/util/statusor.h"
#include "tink/util/validation.h"
#include "proto/common.pb.h"
#include "proto/jwt_hmac.pb.h"
#include "proto/tink.pb.h"

namespace crypto {
namespace tink {
namespace jwt_internal {

using crypto::tink::util::Status;
using crypto::tink::util::StatusOr;
using google::crypto::tink::JwtHmacKey;
using google::crypto::tink::JwtHmacKeyFormat;

uint32_t JwtHmacKeyManager::get_version() const {
  return raw_key_manager_.get_version();
}

google::crypto::tink::KeyData::KeyMaterialType
JwtHmacKeyManager::key_material_type() const {
  return raw_key_manager_.key_material_type();
}

const std::string& JwtHmacKeyManager::get_key_type() const {
  return raw_key_manager_.get_key_type();
}

StatusOr<JwtHmacKey> JwtHmacKeyManager::CreateKey(
    const JwtHmacKeyFormat& jwt_hmac_key_format) const {
  return raw_key_manager_.CreateKey(jwt_hmac_key_format);
}

StatusOr<JwtHmacKey> JwtHmacKeyManager::DeriveKey(
    const JwtHmacKeyFormat& jwt_hmac_key_format,
    InputStream* input_stream) const {
  return raw_key_manager_.DeriveKey(jwt_hmac_key_format, input_stream);
}

Status JwtHmacKeyManager::ValidateKey(const JwtHmacKey& key) const {
  return raw_key_manager_.ValidateKey(key);
}

Status JwtHmacKeyManager::ValidateKeyFormat(
    const JwtHmacKeyFormat& key_format) const {
  return raw_key_manager_.ValidateKeyFormat(key_format);
}

}  // namespace jwt_internal
}  // namespace tink
}  // namespace crypto

File: go/jwt/jwt_hmac_key_manager.go
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package jwt

import (
	"fmt"

	"google.golang.org/protobuf/proto"
	"github.com/google/tink/go/core/registry"
	"github.com/google/tink/go/keyset"
	macsubtle "github.com/google/tink/go/mac/subtle"
	"github.com/google/tink/go/subtle/random"
	"github.com/google/tink/go/subtle"
	jwtmacpb "github.com/google/tink/go/proto/jwt_hmac_go_proto"
	tinkpb "github.com/google/tink/go/proto/tink_go_proto"
)

const (
	jwtHMACKeyVersion = 0
	jwtHMACTypeURL    = "type.googleapis.com/google.crypto.tink.JwtHmacKey"
)

// jwtHMACKeyManager is an implementation of the KeyManager interface
type jwtHMACKeyManager struct{}

// Assert that jwtHMACKeyManager implements the KeyManager interface.
var _ registry.KeyManager = (*jwtHMACKeyManager)(nil)

var hsAlgToHash = map[jwtmacpb.JwtHmacAlgorithm]string{
	jwtmacpb.JwtHmacAlgorithm_HS256: "SHA256",
	jwtmacpb.JwtHmacAlgorithm_HS384: "SHA384",
	jwtmacpb.JwtHmacAlgorithm_HS512: "SHA512",
}

var hsAlgToMinKeySizeBytes = map[jwtmacpb.JwtHmacAlgorithm]int{
	jwtmacpb.JwtHmacAlgorithm_HS256: 32,
	jwtmacpb.JwtHmacAlgorithm_HS384: 48,
	jwtmacpb.JwtHmacAlgorithm_HS512: 64,
}

func (km *jwtHMACKeyManager) Primitive(serializedKey []byte) (interface{}, error) {
	key := &jwtmacpb.JwtHmacKey{}
	if err := proto.Unmarshal(serializedKey, key); err != nil {
		return nil, err
	}
	if err := km.validateKey(key); err != nil {
		return nil, err
	}
	hashAlg, ok := hsAlgToHash[key.GetAlgorithm()]
	if !ok {
		return nil, fmt.Errorf("invalid algorithm: '%v'", key.GetAlgorithm())
	}
	tagSize, err := subtle.GetHashDigestSize(hashAlg)
	if err != nil {
		return nil, err
	}
	mac, err := macsubtle.NewHMAC(hashAlg, key.GetKeyValue(), tagSize)
	if err != nil {
		return nil, err
	}
	var kid *string = nil
	if key.GetCustomKid() != nil {
		k := key.GetCustomKid().GetValue()
		kid = &k
	}
	return newMACWithKID(mac, key.GetAlgorithm().String(), kid)
}

func (km *jwtHMACKeyManager) NewKey(serializedKeyFormat []byte) (proto.Message, error) {
	if serializedKeyFormat == nil || len(serializedKeyFormat) == 0 {
		return nil, fmt.Errorf("no serialized key format")
	}
	keyFormat := &jwtmacpb.JwtHmacKeyFormat{}
	if err := proto.Unmarshal(serializedKeyFormat, keyFormat); err != nil {
		return nil, err
	}
	if err := km.validateKeyFormat(keyFormat); err != nil {
		return nil, err
	}
	return &jwtmacpb.JwtHmacKey{
		Version:   jwtHMACKeyVersion,
		Algorithm: keyFormat.GetAlgorithm(),
		KeyValue:  random.GetRandomBytes(keyFormat.KeySize),
	}, nil
}

func (km *jwtHMACKeyManager) NewKeyData(serializedKeyFormat []byte) (*tinkpb.KeyData, error) {
	key, err := km.NewKey(serializedKeyFormat)
	if err != nil {
		return nil, err
	}
	serializedKey, err := proto.Marshal(key)
	if err != nil {
		return nil, err
	}
	return &tinkpb.KeyData{
		TypeUrl:         jwtHMACTypeURL,
		Value:           serializedKey,
		KeyMaterialType: tinkpb.KeyData_SYMMETRIC,
	}, nil
}

// DoesSupport indicates if this key manager supports the given key type.
func (km *jwtHMACKeyManager) DoesSupport(keyTypeURL string) bool {
	return jwtHMACTypeURL == keyTypeURL
}

// TypeURL returns the key type of keys managed by this key manager.
func (km *jwtHMACKeyManager) TypeURL() string {
	return jwtHMACTypeURL
}

func (km *jwtHMACKeyManager) validateKey(key *jwtmacpb.JwtHmacKey) error {
	if key == nil {
		return fmt.Errorf("key can't be nil")
	}
	if err := keyset.ValidateKeyVersion(key.Version, jwtHMACKeyVersion); err != nil {
		return err
	}
	minKeySizeBytes, ok := hsAlgToMinKeySizeBytes[key.GetAlgorithm()]
	if !ok {
		return fmt.Errorf("invalid algorithm: '%v'", key.GetAlgorithm())
	}
	if len(key.KeyValue) < minKeySizeBytes {
		return fmt.Errorf("invalid JwtHmacKey: KeyValue is too short")
	}
	return nil
}

func (km *jwtHMACKeyManager) validateKeyFormat(keyFormat *jwtmacpb.JwtHmacKeyFormat) error {
	if keyFormat == nil {
		return fmt.Errorf("key format can't be nil")
	}
	minKeySizeBytes, ok := hsAlgToMinKeySizeBytes[keyFormat.GetAlgorithm()]
	if !ok {
		return fmt.Errorf("invalid algorithm: '%v'", keyFormat.GetAlgorithm())
	}
	if int(keyFormat.KeySize) < minKeySizeBytes {
		return fmt.Errorf("invalid JwtHmacKeyFormat: KeySize is too small")
	}
	return nil
}
